File: common/testing/src/main/java/ru/ifmo/se/dating/container/Postgres.java
package ru.ifmo.se.dating.container;

import org.testcontainers.containers.PostgreSQLContainer;

import lombok.ToString;

@ToString
public final class Postgres implements AutoCloseable {
    private static final String DOCKER_IMAGE_NAME = "postgres";

    private final PostgreSQLContainer<?> container;

    private Postgres() {
        this.container = new PostgreSQLContainer<>(DOCKER_IMAGE_NAME);
    }

    public String jdbcUrl() {
        return container.getJdbcUrl();
    }

    public String username() {
        return container.getUsername();
    }

    public String password() {
        return container.getPassword();
    }

    @Override
    public void close() throws Exception {
        container.stop();
        container.close();
    }

    public static Postgres start() {
        final var postgres = new Postgres();
        postgres.container.start();
        return postgres;
    }
}

File: common/testing/src/main/kotlin/ru/ifmo/se/dating/container/Postgres.kt
package ru.ifmo.se.dating.container

import org.testcontainers.containers.PostgreSQLContainer

class Postgres private constructor() : AutoCloseable {
    private val container = PostgreSQLContainer(DOCKER_IMAGE_NAME)

    fun jdbcUrl(): String =
        container.jdbcUrl

    fun username(): String =
        container.username

    fun password(): String =
        container.password

    override fun close() {
        container.stop()
        container.close()
    }

    companion object {
        private const val DOCKER_IMAGE_NAME = "postgres"

        fun start(): Postgres {
            val postgres = Postgres()
            postgres.container.start()
            return postgres
        }
    }
}
